Add aws support

rabunkosar-dd opened the pull request:
A proposed skeleton PR to support AWS resource expansion

rabunkosar-dd committed:
add AWS support for some services

rabunkosar-dd committed:
init maps

rabunkosar-dd committed:
unit tests for ec2

rabunkosar-dd committed:
add elasticache test and bug fix

rabunkosar-dd committed:
Merge branch 'master' into add_aws_support

rabunkosar-dd commented:
@manugarg can you take a look at this if you have time? that would help replace some of our custom code we have internally to supplement cloudprober. Thank you.

manugarg commented:
> @manugarg can you take a look at this if you have time? that would help replace some of our custom code we have internally to supplement cloudprober. Thank you.

Hello Rabun,

Sorry for the long delay. I've started reviewing. I'll try to finish in a couple of days.

Thanks for submitting this.

manugarg commented on internal/rds/aws/elasticache.go:
```diff
@@ -0,0 +1,267 @@
+
+// expand runs equivalent API calls as "aws describe-instances",
+// and is used to populate the cache.
+func (cl *elastiCacheLister) expand(reEvalInterval time.Duration) {
```
This function will benefit from splitting into multiple methods.
 expand {
   ... get cache clusters
   ... get redis replication groups
}

and then you can add more documentation for each method. It will be good to improve expand's comment too, to explain why are doing this.. cache clusters and replication groups.

manugarg commented on internal/rds/aws/elasticache.go:
```diff
@@ -0,0 +1,267 @@
+func (cl *elastiCacheLister) expand(reEvalInterval time.Duration) {
+	cl.l.Infof("elasticaches.expand: expanding AWS targets")
+
+	resCC, err := cl.clusterclient.DescribeCacheClusters(context.TODO(), nil)
+	if err != nil {
+		cl.l.Errorf("elasticaches.expand: error while listing cache clusters: %v", err)
+		return
+	}
+
+	var ids = make([]string, 0)
+	var cacheList = make(map[string]*cacheData)
+
+	ts := time.Now().Unix()
+	for _, r := range resCC.CacheClusters {
```
nits:
I found these variables to be a bit confusing. For example, using 'r' for cache cluster creates some confusion as to what we are looking at.

Also, using 'res' for response here is a bit confusing as 'res' can also mean resource in this context. Can you please use 'resp', instead.

manugarg reviewed (commented):
Thanks once again for sending this change and your patience. I've not finished reviewing, but sending in the comments so far. Any chance you can split this change it into:

1. 1 PR to add AWS targets support.
2. 1 PR to add elasticache and RDS.

rabunkosar-dd committed:
split rds and ec API calls

rabunkosar-dd committed:
remove test files

rabunkosar-dd commented:
> Thanks once again for sending this change and your patience. I've not finished reviewing, but sending in the comments so far. Any chance you can split this change it into:
> 
> 1. 1 PR to add AWS targets support.
> 2. 1 PR to add elasticache and RDS.

I have addressed some of your feedback including the splits for AWS APIs. I'll work on splitting the PR into multiple pieces for easy reviews. Thanks for the reviews so far

rabunkosar-dd committed:
update comments

rabunkosar-dd committed:
remove AWS resource discovery for a separate PR

manugarg commented on internal/rds/aws/proto/config.proto:
```diff
@@ -0,0 +1,114 @@
+option go_package = "github.com/cloudprober/cloudprober/internal/rds/aws/proto";
+
+message EC2Instances {
+  // How often resources should be refreshed.
+  optional int32 re_eval_sec = 98 [default = 600];  // default 10 mins
```
I am assuming you'll eventually need names and filters for all these resources too?

rabunkosar-dd replied:
yes that's correct

manugarg commented on internal/rds/aws/proto/config.proto:
```diff
@@ -0,0 +1,114 @@
+
+// LoadBalancers discovery options.
+message LoadBalancers {
+  // Amazon Resource Name (ARN) of the load balancer
+  // if specified, only the corresponding load balancer information is returned.
+  repeated string name = 1;
```
As name can be slightly confusing, should we standardize on ARN (arn) for AWS resources?

rabunkosar-dd replied:
that was a wrong comment on my end, this was for the name of the LB not ARN, we can ARN later if needed. I plan to add support for LBs as a separate PR and then multiple options (ARN and name etc) can be added as filters if needed

manugarg reviewed (commented):
Sorry for delay again. Please see my review inline.

manugarg commented:
Hi @rabunkosar-dd, do you want to continue working on it.

Apologies for the initials delays in review, but I think you're pretty close.

(P.S. Bigger PRs require blocking more time but if you break it up, we can get through them more quickly)

rabunkosar-dd commented:
> Hi @rabunkosar-dd, do you want to continue working on it.
> 
> Apologies for the initials delays in review, but I think you're pretty close.
> 
> (P.S. Bigger PRs require blocking more time but if you break it up, we can get through them more quickly)

Hey, I'm planning to resume working on it, got blocked on other projects. Will update the PR soon

rabunkosar-dd committed:
address comments

rabunkosar-dd commented on cmd/cloudprober.go:
it's imports tool, placing the internal package in the top section, seems like a good change but I can remove if you want

rabunkosar-dd committed:
update the comment

manugarg commented on internal/rds/aws/aws_test.go:
```diff
@@ -0,0 +1,133 @@
+// Copyright 2019 The Cloudprober Authors.
+//
```
I'd put the latest year in the copyright headers of new *.go files.

manugarg reviewed (commented):
I think it's getting close. Added some comments about dropping irrelevant file:
  - cmd/cloudprober.go
  - targets/endpoint/endpoint.go
  - rds/client/client.go

Also, it will be great if you can trim down the proto file for now.

Please see my comments inline inline as well.

conallob commented:
This PR is almost 12 months old. I'm prepared to clone the development branch and wrap up this PR unless anyone objects.


If you prefer I do not, please say so by 2024-11-28 aka Thanksgiving, when I will clone the development branch and ensure rabunkosar-dd@ is recognised as co-author

fromz commented:
@conallob do you still have a plan to complete this work?

conallob commented:
@fromz Apologies for the delay. Yes, I ended up getting busier than I expected, so I gave @rabunkosar-dd more time.

https://github.com/cloudprober/cloudprober/pull/1075 is now a WIP, as I work through all the merge conflicts between my fork of https://github.com/rabunkosar-dd/cloudprober/tree/add_aws_support and rebase it against https://github.com/cloudprober/cloudprober/tree/main
